returnn/torch/util/gradient_checkpoint.py

```diff
@@ -0,0 +1,577 @@
+        return bool(self.graph_tensor_from_weak_raw_tensor)
+
+    def record_op(self, func: Any, args: Sequence[Any], kwargs: Dict[str, Any], out: Any):
+        """record op"""
+        out_flat, _ = pytree.tree_flatten(out)
+        wrapped_args = pytree.tree_map_only(torch.Tensor, self.maybe_map_raw_tensor_to_graph_tensor, args)
+        wrapped_kwargs = pytree.tree_map_only(torch.Tensor, self.maybe_map_raw_tensor_to_graph_tensor, kwargs)
+        op = _GraphOp(
+            graph=self,
+            func=func,
+            args=wrapped_args,
+            kwargs=wrapped_kwargs,
+            out_flat_num=len(out_flat),
+        )
+        self.ops_to_be_recomputed.append(op)
+        for i, out_flat_elem in enumerate(out_flat):
+            if isinstance(out_flat_elem, torch.Tensor):
+                if out_flat_elem in self.graph_tensor_from_weak_raw_tensor:
+                    continue
+                tensor_ = _GraphTensor(op=op, out_flat_idx=i)
+                self.graph_tensor_from_weak_raw_tensor[out_flat_elem] = tensor_
+
+    def maybe_store_rng_state(self, arg: Any):
+        """store RNG state if not yet stored for this device."""
+        if isinstance(arg, torch.Tensor):
+            device = arg.device
+        elif isinstance(arg, torch.device):
+            device = arg
+        else:
+            return
+        if device not in self.stored_device_rng_states:
+            self.stored_device_rng_states[device] = _get_dev_rng_state(device)
+
+    def maybe_store_amp_state(self, arg: Any):
+        """store AMP state if not yet stored for this device."""
+        if isinstance(arg, torch.Tensor):
+            device = arg.device
+        elif isinstance(arg, torch.device):
+            device = arg
+        else:
+            return
+        if device not in self.stored_device_amp_states:
+            self.stored_device_amp_states[device] = _get_dev_amp_state(device)
+
+    def maybe_map_raw_tensor_to_graph_tensor(self, tensor: torch.Tensor) -> Union[_GraphTensor, torch.Tensor]:
+        """raw tensor to graph tensor if available, otherwise return raw tensor."""
+        return self.graph_tensor_from_weak_raw_tensor.get(tensor, tensor)
+
+    def maybe_recompute(self):
+        """
+        Recompute.
+
+        Make sure that the recomputations happen in the correct order,
+        to get any random number generator state correct.
+
+        Note that we considered to have an API here which allowed to only recompute a subset of the ops.
+        It would still compute all from op idx 0 to some given op idx, but not the rest.
+        On subsequent calls, it would then continue from the last idx until again the requested op idx.
+        This works fine except of one important aspect: The RNG state.
+        If there are any other ops in between which use the RNG state, the RNG state would not be correct anymore.
+        So we cannot allow this. We must recompute all ops together right now.
```

Comment:
Ah yes. But getting and resetting the RNG state also has some overhead (I'm not sure how much, but I assume it's not really optimized to do that often).

> it will only be worth it if we only use a small subset of the recomputed tensors anyway

No, that's not the only reason. It also potentially might require less memory when it is computed more on demand.

I will maybe add a comment for this for a potential future extension. For my main use cases (weight dropout, variational noise), this would anyway not make any difference.